Workaround for NFS bug in kernel

gllghr opened the pull request:
When clients mount subdirectories of NFSv3 exports, invalid entries can
be left behind in the kernel's export cache. These entries have a
reference to the mount containing the path the entry represents,
preventing the mount from being unmounted. We work around this by
immediately expiring the invalid entries from userspace after they have
been added by the kernel.

I'll continue working on getting a fix upstreamed so that we don't have
to maintain this workaround forever.

gllghr commented:
See http://reviews.delphix.com/r/52220/ for testing

gllghr commented:
cc @sebroy @grwilson

sebroy commented on utils/mountd/cache.c:
```diff
@@ -888,7 +888,7 @@ static int dump_to_cache(int f, char *buf, int buflen, char *domain,
 	time_t now = time(0);
 
-	if (ttl <= 1)
+	if (ttl == 0)
```
What's the significance of this change? Can a negative ttl be passed in?

gllghr replied:
Yes, if you want the entry you are adding to be immediately expired. I can add a comment to that effect.

sebroy commented on utils/mountd/cache.c:
```diff
@@ -1429,6 +1429,50 @@ int cache_process_req(fd_set *readfds)
+	char *exp_path = exp->m_export.e_path;
+	char *domain = exp->m_client->m_hostname;
+
+	// Nothing to do if path isn't a subdirectory of exp_path.
```
nit: "//" is not standard C comment-style. Please use "/* bla bla */".

sebroy commented on utils/mountd/cache.c:
```diff
@@ -1429,6 +1429,50 @@ int cache_process_req(fd_set *readfds)
+		char saved = path[l];
+		path[l] = 0;
+		xlog(L_WARNING, "expiring %s\n", path);
+		// A negative TTL will cause the entry to be immediately
```
same here

grwilson commented on utils/mountd/cache.c:
```diff
@@ -881,14 +881,19 @@ static void write_secinfo(char **bp, int *blen, struct exportent *ep, int flag_m
 
 }
 
+/* The ttl parameter represents the amount of time the cache entry will be
```
NIT: they tend to use many different styles for block comments. Is this the most commonly used? I'm used to an empty comment line to start a block and I see they do this too but maybe that's not the common case.

gllghr replied:
I was trying to match the style in the existing code, but looking closer, it doesn't really seem to use a consistent style.

grwilson commented on utils/mountd/cache.c:
```diff
@@ -1429,6 +1435,52 @@ int cache_process_req(fd_set *readfds)
+ *     /a/b/c/d
+ *     /a/b/c/d/e
+ */
+int cache_expire_subdirs(nfs_export *exp, char *path)
```
Will the logic below work if the path passed in is /a/b/c//d/e or /a/b/c/d/e/? I'm not sure if this can happen but wanted to check for those types of corner cases.

gllghr replied:
Good question. The `path` that is passed in was created by a call to `realpath`, so all extraneous slashes should have been removed. Also, the function below is passed the same paths as input and operates on them in a similar fashion, so I think this should be OK regardless.

grwilson replied:
Thanks for verifying.

grwilson commented on utils/mountd/cache.c:
```diff
@@ -1429,6 +1435,52 @@ int cache_process_req(fd_set *readfds)
+		 * A negative TTL will cause the entry to be immediately
+		 * expired.
+		 */
+		dump_to_cache(f, buf, sizeof(buf), domain, path, NULL, -10);
```
Is the -10 TTL value significant here or will any negative value work? If the -10 value is significant then we should document why.

gllghr replied:
There isn't any particular significance to -10. I think anything negative should work, but I picked something more negative than -1 just in case there are edge cases in the expiration/flushing logic in the cache.

grwilson replied:
Ok, maybe drop a comment here so that we indicate that we want a timer that is sufficiently in the past.

gllghr replied:
👍 done

gllghr committed:
Workaround for NFS bug in kernel

When clients mount subdirectories of NFSv3 exports, invalid entries can
be left behind in the kernel's export cache. These entries have a
reference to the mount containing the path the entry represents,
preventing the mount from being unmounted. We work around this by
immediately expiring the invalid entries from userspace after they have
been added by the kernel.
